File: PalindormePartitioningII/PalindromePartitioningIIv1.java
public class PalindromePartitioningIIv1{
    public static void main(String[] args) {
        System.out.println(minCut("aab"));
        System.out.println(minCut("a"));
        System.out.println(minCut("ab"));
        System.out.println(minCut("bb"));
        System.out.println(minCut("abbab"));
    }

    public static int minCut(String s){
        if (s.length() == 1) {return 0;}
        if (isPalindrome(s)) {return ((int) s.length() /2) - 1;}
        return (int) s.length()/2;
    }

    public static boolean isPalindrome(String word){
        word = word.toLowerCase();
        
        int izquierda = 0;
        int derecha = word.length() - 1;

        while (izquierda < derecha) {
            if (word.charAt(izquierda) != word.charAt(derecha)) {
                return false;
            }
            izquierda++;
            derecha--;
        }
        return true;
    }
}

File: PalindormePartitioningII/PalindromePartitioningIIv1.py
def minCut(s):
    n = len(s)
    dp = [0] * n
    for i in range(n):
        dp[i] = i

    # Use center expansion to find all palindromic substrings and update dp
    for center in range(2 * n - 1):
        left = center // 2
        right = left + center % 2

        # Expand while characters are equal
        while left >= 0 and right < n and s[left] == s[right]:
            if left == 0:
                # If the palindrome starts at index 0, no cuts are needed
                dp[right] = 0
            else:
                # Update the minimum number of cuts for this position
                dp[right] = min(dp[right], dp[left - 1] + 1)

            # Move the pointers outward
            left -= 1
            right += 1

    return dp[n - 1]


def main():
    test_cases = ["aab", "a", "ab", "bb", "abbab", "abcba", "abacaba"]

    for s in test_cases:
        print(f'{s} → {minCut(s)}')


# Entry point
if __name__ == "__main__":
    main()
